keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -400,10 +400,12 @@ which would have the following behavior, both for CRD types and built-in types:
 
 #### Non-pointer fields
 
-Non-pointer fields are also important, since they have a specific unmarshalling mechanism in Go.
+Non-pointer fields are also important, since they have a specific
+unmarshalling mechanism in Go.
 
-For CRDs, we can assume that invalid zero-values are omitted, otherwise
-the objects would technically be invalid.
+For CRDs, invalid zero-values are omitted, otherwise the objects would
+technically be invalid. We can safely default these field if they are
+unspecified.
```

Comment:
what does "are omitted" mean? I cannot follow.

Are you saying we are adding a validation step before defaulting for all fields that have zero value?

Note that being invalid is not an easy property to check.

Example:

```yaml
type: string
maxLength: 3
pattern: ^....$
```

or

```
type: string
anyOf:
- minLength: 5
- maxLength: 4
```

In other words, you have to apply the full go-openapi/validator.

Reply:
I'm not planning on changing any of the validation. If we allow empty strings today, we'll continue allowing empty strings with this proposal, but they won't be defaulted. Defaults will only be applied to unspecified or null values on non-nullable fields.

I agree that this doesn't solve all the problems that we would like to solve.

One such problem is for example, container protocol associative key. If someone embeds that container port structure in a CRD, and the "" string is passed for the protocol, it won't be defaulted and will do the wrong thing when it comes to using the associative key (CRD will think the key is `[$port, ""]`, but when the pod template is applied, the empty string will get defaulted to TCP and the key will become `[$port, TCP]`).

To be fair, the fact that we can't differentiate between empty strings and missing strings for built-in type is not addressed by this proposal and still needs to be solved.

Reply:
We can improve incrementally on that in a backward compatible way since this doesn't close the door to improve the zero-value semantics later, if we decide that it is worth it.

Reply:
IMO this is a blocker for the whole enhancement. It clearly states in the goals:

> we will make the behavior equivalent between built-in types and CRDs

This goal is not reached.

Reply:
```
x-kubernetes-prune-zero: true
```

How about something like that? And kubebuilder automatically adding that.

Reply:
Yeah, that's the obvious solution, which would be generated on "omitEmpty" non-pointers I assume. I like that this is not an option to users.

Reply:
Am fine with something more advanced in kubebuilder that know the Go typesystem and when json.Marshal would apply omitempty and when it doesn't. In the CRD schema API we only have to provide the primitives such that kubebuilder can apply them accordingly.